modules/light-clients/08-wasm/types/store.go

```diff
@@ -11,93 +11,150 @@ import (
 
 	errorsmod "cosmossdk.io/errors"
 	"cosmossdk.io/store/cachekv"
-	"cosmossdk.io/store/listenkv"
 	storeprefix "cosmossdk.io/store/prefix"
 	"cosmossdk.io/store/tracekv"
 	storetypes "cosmossdk.io/store/types"
 
 	"github.com/cosmos/ibc-go/v8/modules/core/exported"
 )
 
-// updateProposalWrappedStore combines two KVStores into one while transparently routing the calls based on key prefix
+var (
+	_ wasmvmtypes.KVStore = &storeAdapter{}
+	_ storetypes.KVStore  = &updateProposalWrappedStore{}
+
+	subjectPrefix    = []byte("subject/")
+	substitutePrefix = []byte("substitute/")
+)
+
+// updateProposalWrappedStore combines two KVStores into one.
+//
+// Both stores are used for reads, but only the subjectStore is used for writes. For all operations, the key
+// is checked to determine which store to use and must be prefixed with either "subject/" or "substitute/" accordingly.
+// If the key is not prefixed with either "subject/" or "substitute/", a panic is thrown.
 type updateProposalWrappedStore struct {
 	subjectStore    storetypes.KVStore
 	substituteStore storetypes.KVStore
-
-	subjectPrefix    []byte
-	substitutePrefix []byte
 }
 
-func newUpdateProposalWrappedStore(subjectStore, substituteStore storetypes.KVStore, subjectPrefix, substitutePrefix []byte) updateProposalWrappedStore {
+func newUpdateProposalWrappedStore(subjectStore, substituteStore storetypes.KVStore) updateProposalWrappedStore {
 	return updateProposalWrappedStore{
-		subjectStore:     subjectStore,
-		substituteStore:  substituteStore,
-		subjectPrefix:    subjectPrefix,
-		substitutePrefix: substitutePrefix,
+		subjectStore:    subjectStore,
+		substituteStore: substituteStore,
 	}
 }
 
+// Get implements the storetypes.KVStore interface. It allows reads from both the subjectStore and substituteStore.
+//
+// Get will panic if the key is not prefixed with either "subject/" or "substitute/".
 func (ws updateProposalWrappedStore) Get(key []byte) []byte {
-	return ws.getStore(key).Get(ws.trimPrefix(key))
+	prefix, key := splitPrefix(key)
+
+	return ws.getStore(prefix).Get(key)
 }
 
+// Has implements the storetypes.KVStore interface. It allows reads from both the subjectStore and substituteStore.
+//
+// Note: contracts do not have access to the Has method, it is only implemented here to satisfy the storetypes.KVStore interface.
 func (ws updateProposalWrappedStore) Has(key []byte) bool {
-	return ws.getStore(key).Has(ws.trimPrefix(key))
+	prefix, key := splitPrefix(key)
+
+	return ws.getStore(prefix).Has(key)
 }
 
+// Set implements the storetypes.KVStore interface. It allows writes solely to the subjectStore.
+//
+// Set will panic if the key is not prefixed with "subject/".
 func (ws updateProposalWrappedStore) Set(key, value []byte) {
-	ws.getStore(key).Set(ws.trimPrefix(key), value)
+	prefix, key := splitPrefix(key)
+	if !bytes.Equal(prefix, subjectPrefix) {
+		panic(errors.New("key must be prefixed with subject/"))
```

Comment:
should the panic state that only writes to the subject store are allowed instead?

Reply:
looks great to me!